### PR TITLE
fix: on discount amount calculate marketing expense

### PR DESCRIPTION
**Pull request by vishdha**
TASK ID: https://bloomstack.com/desk#Form/Task/TASK-2020-01054

- [x] Apply 100% discount and based  on discount amount show marketing expense

- [x] Report For Marketing Expense

- [x] On the Customer Dashboard pull the additional Discount amount as Marketing Expense.

### CODE REVIEW

File: erpnext/accounts/report/marketing_expense/marketing_expense.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ "disable_prepared_report": 0,
+ "disabled": 0,
+ "docstatus": 0,
+ "doctype": "Report",
+ "idx": 0,
+ "is_standard": "Yes",
+ "modified": "2020-08-09 00:44:02.499574",
+ "modified_by": "Administrator",
+ "module": "Accounts",
+ "name": "Marketing Expense",
```
Comment (Alchez) [inline erpnext/accounts/report/marketing_expense/marketing_expense.json]: Call this "Marketing Expense**s**"

File: erpnext/accounts/report/marketing_expense/marketing_expense.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ "owner": "Administrator",
+ "prepared_report": 0,
+ "query": "SELECT\n    name as \"ID:Link/Sales Invoice:100\",\n    posting_date as \"Date::100\",\n    customer as \"Customer:Link/Customer:150\",\n    company as \"Company::180\",\n    order_type as \"Order Type::120\",\n    discount_amount as \"Marketing Expense::120\",\n    Status as \"status::200\"\nFROM\n    `tabSales Invoice`\nWHERE\n    order_type = \"Marketing\"\n    AND company = %(company)s\n    AND posting_date BETWEEN %(from_date)s AND %(to_date)s\n",
```
Comment (Alchez) [inline erpnext/accounts/report/marketing_expense/marketing_expense.json]: Don't pull the Order Type field in the report, it will always be Marketing anyway

Comment (Alchez) [inline erpnext/accounts/report/marketing_expense/marketing_expense.json]: Do we want to show values per transaction, or per customer?

**Review by Alchez — CHANGES_REQUESTED** [review]
I don't see any code for setting the discount as 100% when Marketing is selected

File: erpnext/accounts/report/marketing_expense/marketing_expense.json (hunk 2 above)
Comment (vishdha) [reply erpnext/accounts/report/marketing_expense/marketing_expense.json]: I dont know about that will arundhati confirm this

**Comment by vishdha** [comment]
> selected

I can do that

**Comment by vishdha** [comment]
> I don't see any code for setting the discount as 100% when Marketing is selected

OK. Dont know about that but I can add automation

File: erpnext/accounts/report/marketing_expense/marketing_expense.json (hunk 2 above)
Comment (ArundhatiSh) [reply erpnext/accounts/report/marketing_expense/marketing_expense.json]: @Alchez @vishdha I saw the report on Vishal's local and it seems to be ok to have per transaction and we can put a filter for a customer for individual customer total.

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 3]
```diff
@@ -820,6 +820,12 @@ frappe.ui.form.on('Sales Invoice', {
 			method: "erpnext.accounts.doctype.sales_invoice.sales_invoice.create_invoice_discounting",
 			frm: frm
 		});
+	},
+	order_type: function(frm) {
+		let discount_percentage = 100;
+		if(frm.doc.order_type === "Marketing") {
+			frm.set_value("additional_discount_percentage", discount_percentage);
+		}
 	}
```
Comment (Alchez) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js]: If we want to apply this for all doctypes, put this in the `sales_common.js` file instead

Also, handle the case where the percentage goes back to 0 if something else is selected (check the Sample order logic)

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/selling/sales_common.js [hunk 4]
```diff
@@ -463,24 +463,50 @@ erpnext.selling.SellingController = erpnext.TransactionController.extend({
...
+						}
 					}
-				}
-			})
+				})
+			}
 		}
+	},
+
+	order_type: function(frm) {
```
Comment (Alchez) [inline erpnext/selling/sales_common.js]: This will override any `order_type` calls in individual doctypes. Add a `_super` call to ensure that doesn't happen.

```suggestion
	order_type: function(frm) {
		this._super(frm);
```

File: erpnext/selling/sales_common.js [hunk 5]
```diff
@@ -534,24 +534,51 @@ erpnext.selling.SellingController = erpnext.TransactionController.extend({
...
+		this._super(frm);
+		if(in_list(["Quotation", "Sales Order", "Sales Invoice", "Delivery Note"], frm.doc.doctype)) {
+			if(frm.doc.order_type) {
+				frm.trigger("set_expense_discount");
+			}
 		}
+	},
```
Comment (Alchez) [inline erpnext/selling/sales_common.js]: Add this to the existing selling controller in this file, don't create a new one

**Review by Alchez — CHANGES_REQUESTED** [review]
The query just needs some minor tweaks:

```sql
SELECT
    name as "ID:Link/Sales Invoice:100",
    posting_date as "Date::100",  -- change the label to Posting Date, and make this a date field
    customer as "Customer:Link/Customer:150",
    company as "Company::180",  -- make this a link field
    discount_amount as "Marketing Expense::120",  -- make this a currency field
    status as "Status::200"
FROM
    `tabSales Invoice`
WHERE
    order_type = "Marketing"
    AND company = %(company)s
    AND posting_date BETWEEN %(from_date)s AND %(to_date)s
```

**Comment by Alchez** [comment]
@vishdha, the report looks good, but I'm still seeing $0 on the customer dashboard. Can you look into that?

**Comment by vishdha** [comment]
> @vishdha, the report looks good, but I'm still seeing $0 on the customer dashboard. Can you look into that?

@Alchez Please Review now

File: erpnext/selling/sales_common.js [hunk 6]
```diff
@@ -151,9 +151,27 @@ erpnext.selling.SellingController = erpnext.TransactionController.extend({
...
+	set_expense_discount: function(frm) {
+		let me = this;
+		let percentage_discount = 0;
+
+		if (me.frm.doc.order_type === "Marketing") {
+			me.frm.set_value("apply_discount_on", "Grand Total");
+			percentage_discount = 100;
+		}
+
+		me.frm.set_value("additional_discount_percentage", percentage_discount);
```
Comment (Alchez) [inline erpnext/selling/sales_common.js:167]: Do this on the server side as well (on save), I was able to start an invoice on "Marketing", but it didn't apply the discount because there was no change in the Order Type

Comment (vishdha) [reply erpnext/selling/sales_common.js:167]: @Alchez  Server side check added

File: erpnext/controllers/selling_controller.py [hunk 7]
```diff
@@ -417,6 +418,18 @@ def validate_items(self):
 		from erpnext.controllers.buying_controller import validate_item_type
 		validate_item_type(self, "is_sales_item", "sales")
 
+	def validate_marketing_expense(self):
+		# set discount 100% for marketing order
+		percentage_discount = 0
+		if self.doctype in ("Delivery Note", "Sales Invoice", "Sales Order", "Quotation"):
```
Comment (Alchez) [inline erpnext/controllers/selling_controller.py]: Remove this line, you shouldn't have to hardcode this for classes

File: erpnext/controllers/selling_controller.py [hunk 8]
```diff
@@ -417,6 +418,18 @@ def validate_items(self):
...
+			if self.order_type == "Marketing":
+				frappe.db.set_value(self.doctype, self.name, "apply_discount_on", "Grand Total")
+				percentage_discount = 100
+			frappe.db.set_value(self.doctype, self.name, "additional_discount_percentage", percentage_discount)
```
Comment (Alchez) [inline erpnext/controllers/selling_controller.py]:
```suggestion
		percentage_discount = 0
		if hasattr(self, "order_type") and self.order_type == "Marketing":
			self.apply_discount_on = "Grand Total"
			percentage_discount = 100
		self.additional_discount_percentage = percentage_discount
```

File: erpnext/controllers/selling_controller.py [hunk 9]
```diff
@@ -417,6 +418,18 @@ def validate_items(self):
...
+			if percentage_discount:
+				frappe.msgprint(_("{0}% discount applied").format(percentage_discount))
```
Comment (Alchez) [inline erpnext/controllers/selling_controller.py]: It shouldn't be a dialog, a simple alert should be fine. If we can't do that, then remove this entirely.

File: erpnext/controllers/selling_controller.py [hunk 10]
```diff
@@ -417,6 +418,18 @@ def validate_items(self):
...
+			self.reload()
```
Comment (Alchez) [inline erpnext/controllers/selling_controller.py]: Why is this needed? Aren't we running this in validate?

**Review by Alchez — CHANGES_REQUESTED** [review]

**Review by Alchez — APPROVED** [review]
